### DISCUSSION TITLE
React + Vite pre-rendering pages - doing SSG without SSR

### DISCUSSION THREAD

**Discussion by CookedApps** [vikejs/vike#1117] [Discussion]
Hey guys, 

We've been looking into SSG lately as we want to improve SEO on our website. Most of our website is highly dynamic (it's a browser game) and should stay a client rendered SPA, but some pages (like the landing page) are static and could be pre-renderd at build time. We want to pre-render the static pages in the different translations we have and serve them under different paths. 

vite-plugin-ssr offers SSG functionality, but we couldn't figure out how to add pre-rendering without also implementing SSR to our project. **Our goal** is to use vite-plugin-ssr at build time to pre-generate some specific pages and bundle it together with the SPA part of our website and then deploy it to a static hosting provider (Cloudflare Pages).

Our current setup:
- React 17 SPA
- Vite build system
- Static web hosting on Cloudflare Pages
- This is how our bundle looks like:
```
index.html (entry point, hydrated by JS)
/assets
  /js
    bundle-as238fha.js (Large JS bundle with all app code)
    ...
  /img
  ...
```

What we want:
- Keep most of the SPA stuff (our webiste is heavly dynamic)
- Pre-render some static pages with SSG
- Deploy the bundle of the pre-rendered pages + the SPA stuff to Cloudflare Pages
- Don't use any SSR
- This is how we imagine our bundle with pre-rendering:
```
index.html (Startpage & entry point in English)
/faq
  index.html (FAQ in English)
/tos
  index.html (TOS in English)
...
/de
  index.html (Homepage in German)
  /faq (FAQ in German)
  /tos
  ...
/pl
  index.html (Homepage in Polish)
  /faq
  /tos
  ...
/assets
  /js
  /img
  ...
  ```
  
  How can we achieve this?
  
 We are new to the topic and couldn't find any examples using the same setup, explaining how to pull this off. So it would be great to get some input and help from the community. :)

**Comment by brillout** [comment]
To achieve that, see:
 - https://vite-plugin-ssr.com/render-modes
 - https://vite-plugin-ssr.com/file-structure#domain-driven (so that you can apply your SPA/SSG mode to different pages)
 - https://vite-plugin-ssr.com/i18n

It requires some manual work on your part though. If you don't want to do that manual work, then check out [vike-react](https://github.com/vikejs/vike-react) which features an `ssr: false` flag (see example) to opt-out of SSR. Vike-react is in beta but it's fairly stable already and issues are fixed promptly.

Also keep in mind that: 91e815ce0d0252bc1b456d40b88d93ac8a033783.

I agree an example would be neat. Would you be up for contributing one? Ideally using vike-react, as it would show how easy it is. (Although i18n still requires manual work until https://github.com/brillout/vite-plugin-ssr/issues/1000 is implemented.)

**Reply by CookedApps** [reply]
Awesome, thanks for the detailed answer @brillout! 
We'll look into it, and sure, we can provide an example, once we have something working.

**Reply by brillout** [reply]
👍 Btw: I don't know how much you make with your website (seems like a sound business model 👀), but FYI https://github.com/sponsors/brillout is making a big difference.

**Reply by CookedApps** [reply]
I'll come back to it when you help us sort this out. 😉

**Reply by brillout** [reply]
Yea it's definitely premature 🙂, but looking forward to resume the discussion later on.

**Comment by NimiFlaisher** [comment]
Hi Rom!

I've tried countless times to make pre-rendering work but keep failing on the setup.

Is it possible, perhaps, for you to a short explanation of the initial setup? A step-by-step setup guide? Or anything near that, from creating a project to actually being able to pretender some files?

I'm trying to create an example project, that creates/prerenders a page (for example, FAQ) in multiple languages.

Thanks!

**Reply by brillout** [reply]
What are your issues?

Have you had a look at `examples/i18n` which does pre-rederning?

**Reply by NimiFlaisher** [reply]
I have yet to use Internationalization since I didn't get to the point where I managed to pre-render any files.

About the issues, it's an endless chain of errors in the terminal. Whenever I try to run the build I get new errors about the renderer, the exports, missing files, etc.

I must be doing something wrong and I apologize if I'm being unclear, but it's really hard to explain since I can barely get it to run at all.

If you have any guide showing the initial setup (the installs, all the required files), that would be sufficient.

**Reply by brillout** [reply]
I understand things are difficult to get started with: vite-plugin-ssr has been designed for lower-level control which means more integration work.

I don't think there is any alternative other than for you to dig more and have a better understanding of how a typical vite-plugin-ssr app works.

If you want an easier getting started, then check out [vike-react](https://github.com/vikejs/vike-react) (which, for example, features an `ssr: boolean` config) and its official scaffolder [Bati](https://batijs.github.io/). It's going to get out of beta soon.

FYI I do agree that writing guides would help a lot, but our focus right now is on making getting started easier with things like `vike-react` and Bati.

**Reply by kripainc** [reply]
Have a look at the examples/react-full demo application where pages/start-wars has a simple SSG working example.
Just clone the VPS, cd examples/react-full; npm install; npm run build.
In vite config, the ssr({ prerender: true }) makes it a SSG.

**Reply by NimiFlaisher** [reply]
Thanks kripainc, I actually managed to do that earlier this week, but I failed to implement those on my project, still working to figure it out :)
